app/oidc/oidc.py

```diff
@@ -0,0 +1,66 @@
+import functools
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from typing import Callable, ParamSpec
+
+import pytz
+from google.auth.credentials import Credentials
+from google.auth.transport.requests import Request
+from structlog import get_logger
+
+from app.settings import OIDC_TOKEN_LEEWAY_IN_SECONDS
+
+P = ParamSpec("P")
+
+logger = get_logger()
+TTL = 3600 - OIDC_TOKEN_LEEWAY_IN_SECONDS  # 1 hour minus leeway in seconds
+
+
+def get_expiry_from_ttl(ttl: int) -> datetime:
+    return datetime.now(timezone.utc) - timedelta(seconds=ttl)
+
+
+def is_naive(dt: datetime) -> bool:
+    return dt.tzinfo is None or dt.tzinfo.utcoffset(dt) is None
+
+
+def refresh_oidc_credentials(
+    func: Callable[P, Credentials]
+) -> Callable[P, Credentials]:
+    """
+    This decorator is used on implementations of the abstract method OIDCCredentialsService.get_credentials.
+    This decorated method uses a ttl_cache decorator to discard fetched credentials after a given time.
+    However, the TTL of the credentials should be determined by inspecting the expiry of the credentials
+     themselves, not relying on the ttl_cache.
+    As such, this decorator ensures the credentials returned from the ttl_cache are valid and refreshed
+     via inspecting the expiry property of the credentials.
+    Overall this ensures the caller is always returned valid credentials and credentials are cached to minimise
+     API calls for credentials as well as discarded once invalid.
+    """
+
+    @functools.wraps(func)
+    def wrapper_refresh_oidc_credentials(
+        *args: P.args, **kwargs: P.kwargs
+    ) -> Credentials:
+        credentials = func(*args, **kwargs)
+        expiry = (
+            pytz.utc.localize(dt=credentials.expiry)
+            if is_naive(credentials.expiry)
+            else credentials.expiry
+        )
+        if expiry < get_expiry_from_ttl(TTL):
+            logger.info("refreshing oidc credentials")
```

Comment:
Worth adding the client id.

Reply:
The way we'd get the client id in the context of the decorator would be to decode the token, at that point I'm not sure that's worthwhile?

Reply:
Have defaulted SDS_OAUTH2_CLIENT_ID to None to address another comment, so will just use that, in local the log will show None which isn't really an issue

Reply:
Done 👍

Reply:
Doesn't look like it, it is binding the `SDS_OAUTH2_CLIENT_ID` explicit which won;t be true when the same thing is used for CIR. Just pass `args` and `kwargs` to the logger info, actually. As `kwargs` will contain the iap client id in the current case. Either pass both, or enforce one of positional / keyword only arguments.

Reply:
Binding kwargs to the log 👍